File: src/infrastructure/persistence/spike_follow_up_adapter.py
"""SpikeFollowUpAdapter -- audits spike follow-ups against created tickets.

Implements SpikeFollowUpPort by scanning research reports for follow-up
intents and comparing them against beads tickets using fuzzy title matching.
"""

from __future__ import annotations

import json
import re
from typing import TYPE_CHECKING

from src.domain.models.follow_up_intent import FollowUpAuditResult, FollowUpIntent
from src.infrastructure.persistence.markdown_spike_parser import MarkdownSpikeReportParser

if TYPE_CHECKING:
    from pathlib import Path

# Prefixes stripped before comparison (case-insensitive)
_STRIP_PREFIXES = (
    "task:",
    "spike:",
    "bug:",
    "feature:",
    "(optional)",
)

# Short stop-words excluded from keyword overlap scoring
_STOP_WORDS = frozenset(
    {
        "a",
        "an",
        "the",
        "and",
        "or",
        "of",
        "to",
        "in",
        "for",
        "on",
        "with",
        "is",
        "it",
        "be",
        "as",
        "at",
        "by",
        "from",
        "that",
        "this",
    }
)

# Minimum Jaccard similarity for keyword overlap match
_KEYWORD_OVERLAP_THRESHOLD = 0.4


class SpikeFollowUpAdapter:
    """Implements SpikeFollowUpPort backed by filesystem scanning.

    Locates spike research reports in docs/research/, parses them for
    follow-up intents, and compares against .beads/issues.jsonl.
    """

    def __init__(self) -> None:
        self._parser = MarkdownSpikeReportParser()

    def audit(self, spike_id: str, project_dir: Path) -> FollowUpAuditResult:
        """Audit a spike's follow-up intents against created tickets."""
        research_dir = project_dir / "docs" / "research"
        if not research_dir.exists():
            return FollowUpAuditResult(
                spike_id=spike_id,
                report_path="",
                defined_intents=(),
                matched_ticket_ids=(),
                orphaned_intents=(),
            )

        # Scan all Markdown reports in docs/research/
        all_intents: list[FollowUpIntent] = []
        report_path = ""
        for report_file in sorted(research_dir.glob("*.md")):
            intents = self._parser.parse(report_file)
            if intents:
                all_intents.extend(intents)
                if not report_path:
                    report_path = str(report_file)

        if not all_intents:
            return FollowUpAuditResult(
                spike_id=spike_id,
                report_path=report_path,
                defined_intents=(),
                matched_ticket_ids=(),
                orphaned_intents=(),
            )

        # Load existing beads tickets
        existing_titles = self._load_ticket_titles(project_dir)

        # Match intents against tickets
        matched_ids: list[str] = []
        orphaned: list[FollowUpIntent] = []

        for intent in all_intents:
            ticket_id = self._fuzzy_match(intent.title, existing_titles)
            if ticket_id:
                matched_ids.append(ticket_id)
            else:
                orphaned.append(intent)

        return FollowUpAuditResult(
            spike_id=spike_id,
            report_path=report_path,
            defined_intents=tuple(all_intents),
            matched_ticket_ids=tuple(matched_ids),
            orphaned_intents=tuple(orphaned),
        )

    def _load_ticket_titles(self, project_dir: Path) -> dict[str, str]:
        """Load ticket ID → title mapping from .beads/issues.jsonl."""
        issues_path = project_dir / ".beads" / "issues.jsonl"
        if not issues_path.exists():
            return {}

        titles: dict[str, str] = {}
        for line in issues_path.read_text().splitlines():
            line = line.strip()
            if not line:
                continue
            try:
                issue = json.loads(line)
                ticket_id = issue.get("id", "")
                title = issue.get("title", "")
                if ticket_id and title:
                    titles[ticket_id] = title
            except json.JSONDecodeError:
                continue
        return titles

    def _fuzzy_match(self, intent_title: str, existing: dict[str, str]) -> str | None:
        """Find a beads ticket that fuzzy-matches the intent title.

        Matching strategy (ordered by strictness):
        1. Case-insensitive exact match (after prefix stripping)
        2. Case-insensitive substring (intent in ticket or ticket in intent)
        3. Keyword overlap (Jaccard similarity >= threshold)

        Returns the matching ticket ID, or None.
        """
        stripped_intent = self._strip_prefixes(intent_title)
        if not stripped_intent:
            return None

        intent_lower = stripped_intent.lower()

        for ticket_id, title in existing.items():
            stripped_ticket = self._strip_prefixes(title)
            if not stripped_ticket:
                continue
            ticket_lower = stripped_ticket.lower()

            # Tier 1: Exact match
            if intent_lower == ticket_lower:
                return ticket_id

            # Tier 2: Substring
            if intent_lower in ticket_lower or ticket_lower in intent_lower:
                return ticket_id

            # Tier 3: Keyword overlap
            if self._keyword_overlap(intent_lower, ticket_lower) >= _KEYWORD_OVERLAP_THRESHOLD:
                return ticket_id

        return None

    @staticmethod
    def _strip_prefixes(title: str) -> str:
        """Remove common prefixes from a title for comparison."""
        result = title.strip()
        lower = result.lower()
        for prefix in _STRIP_PREFIXES:
            if lower.startswith(prefix):
                result = result[len(prefix) :].strip()
                lower = result.lower()
        return result

    @staticmethod
    def _tokenize(text: str) -> set[str]:
        """Tokenize a title into meaningful keywords.

        Splits on whitespace/punctuation, lowercases, removes stop-words,
        and extracts parenthetical content as additional tokens.
        """
        # Extract parenthetical content as extra tokens
        parens = re.findall(r"\(([^)]+)\)", text)
        paren_tokens: set[str] = set()
        for p in parens:
            for token in re.split(r"[\s,+/]+", p.lower()):
                token = token.strip()
                if token and token not in _STOP_WORDS and len(token) > 1:
                    paren_tokens.add(token)

        # Main tokenization
        words = re.split(r"[\s\-_/,()]+", text.lower())
        tokens = {w for w in words if w and w not in _STOP_WORDS and len(w) > 1}

        return tokens | paren_tokens

    @classmethod
    def _keyword_overlap(cls, a: str, b: str) -> float:
        """Compute fuzzy keyword overlap between two titles.

        Uses prefix matching (first 5 chars) for each token pair to handle
        morphological variants like "generate"/"generation", "function"/"functions".
        Returns a score from 0.0 to 1.0.
        """
        tokens_a = cls._tokenize(a)
        tokens_b = cls._tokenize(b)
        if not tokens_a or not tokens_b:
            return 0.0

        # Count how many tokens in A have a fuzzy match in B
        matched_a = 0
        matched_b_tokens: set[str] = set()
        for ta in tokens_a:
            for tb in tokens_b:
                if tb in matched_b_tokens:
                    continue
                if ta == tb or (len(ta) >= 5 and len(tb) >= 5 and ta[:5] == tb[:5]):
                    matched_a += 1
                    matched_b_tokens.add(tb)
                    break

        total_unique = len(tokens_a) + len(tokens_b) - matched_a
        if total_unique == 0:
            return 0.0
        return matched_a / total_unique

File: internal/research/infrastructure/spike_follow_up_adapter.go
package infrastructure

import (
	"context"
	"encoding/json"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	researchapp "github.com/alto-cli/alto/internal/research/application"
	researchdomain "github.com/alto-cli/alto/internal/research/domain"
)

// Compile-time interface check.
var _ researchapp.SpikeFollowUp = (*SpikeFollowUpAdapter)(nil)

// Prefixes stripped before comparison (case-insensitive).
var stripPrefixes = []string{
	"task:",
	"spike:",
	"bug:",
	"feature:",
	"(optional)",
}

// Short stop-words excluded from keyword overlap scoring.
var stopWords = map[string]bool{
	"a": true, "an": true, "the": true, "and": true, "or": true,
	"of": true, "to": true, "in": true, "for": true, "on": true,
	"with": true, "is": true, "it": true, "be": true, "as": true,
	"at": true, "by": true, "from": true, "that": true, "this": true,
}

// Minimum Jaccard similarity for keyword overlap match.
const keywordOverlapThreshold = 0.4

// parenRE extracts content from parentheses.
var parenRE = regexp.MustCompile(`\(([^)]+)\)`)

// tokenSplitRE splits on whitespace/punctuation for tokenization.
var tokenSplitRE = regexp.MustCompile(`[\s\-_/,()]+`)

// parenTokenSplitRE splits parenthetical content.
var parenTokenSplitRE = regexp.MustCompile(`[\s,+/]+`)

// SpikeFollowUpAdapter audits spike follow-ups against created tickets.
// Implements SpikeFollowUp port by scanning research reports for follow-up
// intents and comparing them against beads tickets using fuzzy title matching.
type SpikeFollowUpAdapter struct {
	parser *MarkdownSpikeParser
}

// NewSpikeFollowUpAdapter creates a new SpikeFollowUpAdapter.
func NewSpikeFollowUpAdapter() *SpikeFollowUpAdapter {
	return &SpikeFollowUpAdapter{
		parser: NewMarkdownSpikeParser(),
	}
}

// Audit audits a spike's follow-up intents against created tickets.
func (a *SpikeFollowUpAdapter) Audit(ctx context.Context, spikeID string, projectDir string) (researchdomain.FollowUpAuditResult, error) {
	researchDir := filepath.Join(projectDir, "docs", "research")
	if _, err := os.Stat(researchDir); os.IsNotExist(err) {
		return researchdomain.NewFollowUpAuditResult(spikeID, "", nil, nil, nil), nil
	}

	// Scan all Markdown reports in docs/research/
	entries, err := os.ReadDir(researchDir)
	if err != nil {
		return researchdomain.NewFollowUpAuditResult(spikeID, "", nil, nil, nil), nil
	}

	// Sort entries by name for deterministic order
	sort.Slice(entries, func(i, j int) bool {
		return entries[i].Name() < entries[j].Name()
	})

	var allIntents []researchdomain.FollowUpIntent
	reportPath := ""
	for _, entry := range entries {
		if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".md") {
			continue
		}
		fullPath := filepath.Join(researchDir, entry.Name())
		intents, parseErr := a.parser.Parse(ctx, fullPath)
		if parseErr != nil {
			continue
		}
		if len(intents) > 0 {
			allIntents = append(allIntents, intents...)
			if reportPath == "" {
				reportPath = fullPath
			}
		}
	}

	if len(allIntents) == 0 {
		return researchdomain.NewFollowUpAuditResult(spikeID, reportPath, nil, nil, nil), nil
	}

	// Load existing beads tickets
	existingTitles := loadTicketTitles(projectDir)

	// Match intents against tickets
	var matchedIDs []string
	var orphaned []researchdomain.FollowUpIntent

	for _, intent := range allIntents {
		ticketID := FuzzyMatch(intent.Title(), existingTitles)
		if ticketID != "" {
			matchedIDs = append(matchedIDs, ticketID)
		} else {
			orphaned = append(orphaned, intent)
		}
	}

	return researchdomain.NewFollowUpAuditResult(spikeID, reportPath, allIntents, matchedIDs, orphaned), nil
}

// loadTicketTitles loads ticket ID -> title mapping from .beads/issues.jsonl.
func loadTicketTitles(projectDir string) map[string]string {
	issuesPath := filepath.Join(projectDir, ".beads", "issues.jsonl")
	data, err := os.ReadFile(issuesPath)
	if err != nil {
		return map[string]string{}
	}

	titles := make(map[string]string)
	for _, line := range strings.Split(string(data), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		var issue map[string]string
		if err := json.Unmarshal([]byte(line), &issue); err != nil {
			continue
		}
		ticketID := issue["id"]
		title := issue["title"]
		if ticketID != "" && title != "" {
			titles[ticketID] = title
		}
	}
	return titles
}

// FuzzyMatch finds a beads ticket that fuzzy-matches the intent title.
// Exported for testing. Returns the matching ticket ID, or empty string.
//
// Matching strategy (ordered by strictness):
//  1. Case-insensitive exact match (after prefix stripping)
//  2. Case-insensitive substring (intent in ticket or ticket in intent)
//  3. Keyword overlap (Jaccard similarity >= threshold)
func FuzzyMatch(intentTitle string, existing map[string]string) string {
	strippedIntent := stripPrefixesFromTitle(intentTitle)
	if strippedIntent == "" {
		return ""
	}
	intentLower := strings.ToLower(strippedIntent)

	for ticketID, title := range existing {
		strippedTicket := stripPrefixesFromTitle(title)
		if strippedTicket == "" {
			continue
		}
		ticketLower := strings.ToLower(strippedTicket)

		// Tier 1: Exact match
		if intentLower == ticketLower {
			return ticketID
		}

		// Tier 2: Substring
		if strings.Contains(ticketLower, intentLower) || strings.Contains(intentLower, ticketLower) {
			return ticketID
		}

		// Tier 3: Keyword overlap
		if keywordOverlap(intentLower, ticketLower) >= keywordOverlapThreshold {
			return ticketID
		}
	}

	return ""
}

// stripPrefixesFromTitle removes common prefixes from a title for comparison.
func stripPrefixesFromTitle(title string) string {
	result := strings.TrimSpace(title)
	lower := strings.ToLower(result)
	for _, prefix := range stripPrefixes {
		if strings.HasPrefix(lower, prefix) {
			result = strings.TrimSpace(result[len(prefix):])
			lower = strings.ToLower(result)
		}
	}
	return result
}

// tokenize splits a title into meaningful keywords.
// Splits on whitespace/punctuation, lowercases, removes stop-words,
// and extracts parenthetical content as additional tokens.
func tokenize(text string) map[string]bool {
	// Extract parenthetical content as extra tokens
	parenMatches := parenRE.FindAllStringSubmatch(text, -1)
	parenTokens := make(map[string]bool)
	for _, m := range parenMatches {
		for _, token := range parenTokenSplitRE.Split(strings.ToLower(m[1]), -1) {
			token = strings.TrimSpace(token)
			if token != "" && !stopWords[token] && len(token) > 1 {
				parenTokens[token] = true
			}
		}
	}

	// Main tokenization
	words := tokenSplitRE.Split(strings.ToLower(text), -1)
	tokens := make(map[string]bool)
	for _, w := range words {
		if w != "" && !stopWords[w] && len(w) > 1 {
			tokens[w] = true
		}
	}

	// Merge
	for k := range parenTokens {
		tokens[k] = true
	}
	return tokens
}

// keywordOverlap computes fuzzy keyword overlap between two titles.
// Uses prefix matching (first 5 chars) for each token pair to handle
// morphological variants like "generate"/"generation".
// Returns a score from 0.0 to 1.0.
func keywordOverlap(a, b string) float64 {
	tokensA := tokenize(a)
	tokensB := tokenize(b)
	if len(tokensA) == 0 || len(tokensB) == 0 {
		return 0.0
	}

	matchedA := 0
	matchedBTokens := make(map[string]bool)
	for ta := range tokensA {
		for tb := range tokensB {
			if matchedBTokens[tb] {
				continue
			}
			if ta == tb || (len(ta) >= 5 && len(tb) >= 5 && ta[:5] == tb[:5]) {
				matchedA++
				matchedBTokens[tb] = true
				break
			}
		}
	}

	totalUnique := len(tokensA) + len(tokensB) - matchedA
	if totalUnique == 0 {
		return 0.0
	}
	return float64(matchedA) / float64(totalUnique)
}
